### DISCUSSION TITLE
Is there a way to have a install_crate depedency for init tasks/plugins

### DISCUSSION THREAD

**Discussion by wmmc88** [sagiegurari/cargo-make#982] [Q&A] [answered]
I have an init task that uses a cargo-make plugin to invoke rust-script. Is there a way to enforce that rust-script is actually insatlled (like `install_crate = { crate_name = "rust-script", min_version = "0.30.0" }`)? It seems that either `install_crate` doesnt work for init tasks, or it doesnt get executed before the plugin.

**Comment by wmmc88** [comment]
On second thought, is it because for tasks that specify plugins, you're supposed to handle the execution yourself, so it doesn't actually do anything with the `install_crate`? Is the only way to parse that `install_crate` section myself in the plugin code and execute it?

**Reply by sagiegurari** [reply]
can you share a very simple basic example for me to look at?

**Reply by wmmc88** [reply]
My actual use case is here: https://github.com/microsoft/windows-drivers-rs/pull/35/commits/4cb9220233c7898bf5fabc88b9d1746885dd8b25#diff-5281732389c23b9261ec5548106166183e9bb899f98857736f14520410519575. I want my `rust-env-update` plugin to execute via my `wdk-build-init` task, but the plugin has a dependency on rust-script being installed. putting `install_crate` in the task doesnt work, and I explictly added logic in the plugin to handle the installation of rust-script.

**For a more minimal example:**

makefile.toml
```
[tasks.task-a]
install_crate = { crate_name = "rust-script", min_version = "0.30.0" }
plugin = "plugin-a"

[plugins.impl.plugin-a]
script = '''
out = exec --fail-on-error rust-script --help
assert_eq ${out.code} 0
'''
```

```
🪟 Windows 11 Enterprise(10.0.23580) (pwsh) D:\git-repos took 4s
at 12:16:59 PM ❯ cargo uninstall rust-script
    Removing D:\.tools\.cargo\bin\rust-script.exe

🪟 Windows 11 Enterprise(10.0.23580) (pwsh) D:\git-repos
at 12:17:10 PM ❯ cargo make task-a
[cargo-make] INFO - cargo make 0.37.4
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: task-a
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: legacy-migration
[cargo-make] ERROR - Error while running plugin: Source: Unknown Line: 5 - program not found
[cargo-make] WARN - Build Failed.
```

Unrelated to this discussion, but I've also found that cargo-make doesn't give particularily useful error message when failing duckscript. The line numbers seem to not line up to the script blocks, and it would be nice that if in examples like above, "program not found" could be more descriptive and specify what "program" was not found

**Reply by sagiegurari** [reply]
@wmmc88 sorry for late answer
1. line numbers - right, its because i put some extra code before yours. for example to exit with error on any error as by default duckscript will not stop on error (like shell scripts don't by default)
2. getting the install crate info - 
* task.as_json gives you the entire task data as json
so you can do 
taskjson = json_parse ${task.as_json}
# holds the data -> taskjson.install_crate
* there are other vars that help. see https://github.com/sagiegurari/cargo-make#plugin-sdk

**Reply by sagiegurari** [reply]
@wmmc88 did it answer you?

**Reply by wmmc88** [reply]
yes, I understand how it works now. Thank you!

It would be nice if line numbers were synced to actual lines in the code, or there was a way to match the error line to the script line somehow(Does the duckscript get saved in a file somewhere where the lines would match?).

It would also be nice if there was a way to defer to the default behaviour somehow in the plugin. For example, sometimes I want to have a plugin that just adds a little extra processing, but still want all the nice things that cargo make provides for me (dependency installation, min version checking, etc.) without needing to reimplement every single feature in the duckscript for the plugin

**Reply by sagiegurari** [reply, marked as answer]
> Does the duckscript get saved in a file

no. since its embedded inside cargo-make, the script is in memory....
i could dump it via verbose to outside if that helps (maybe i already do that, can't remember)

> It would also be nice if there was a way to defer to the default behaviour somehow in the plugin

cm_plugin_run_custom_task enables you to modify the task and run it.
it handles all logic except dependencies, because dependencies should have already been executed so i don't want to execute them again. well, unless i forgot how it works lol
